src/ode_solver/pod_galerkin_ode_solver.cpp

```diff
@@ -36,39 +119,106 @@ void PODGalerkinODESolver<dim,real,MeshType>::step_in_time (real dt, const bool
+    Epetra_LinearProblem linearProblem(&epetra_reduced_lhs, &epetra_reduced_solution_update, &epetra_reduced_rhs);
+
+    Amesos Factory;
+    std::string SolverType = "Klu";
+    std::unique_ptr<Amesos_BaseSolver> Solver(Factory.Create(SolverType, linearProblem));
+
+    Teuchos::ParameterList List;
+    Solver->SetParameters(List);
+    Solver->SymbolicFactorization();
+    Solver->NumericFactorization();
+    Solver->Solve();
+
+    this->pcout << "Reduced solution update norm: " << reduced_solution_update.l2_norm() << std::endl;
+
+    const dealii::LinearAlgebra::distributed::Vector<double> old_reduced_solution(reduced_solution);
+    double step_length = 1.0;
+    const double step_reduction = 0.5;
+    const int maxline = 10;
+    const double reduction_tolerance_1 = 1.0;
+
+    double initial_residual;
+    epetra_reduced_rhs.Norm2(&initial_residual);
+    initial_residual /= this->dg->right_hand_side.size();
+
+    reduced_solution.add(step_length, this->reduced_solution_update);
+    Epetra_Vector epetra_reduced_solution(Epetra_DataAccess::View, epetra_pod_basis.DomainMap(), reduced_solution.begin());
+    Epetra_Vector solution(Epetra_DataAccess::View, epetra_pod_basis.RangeMap(), this->dg->solution.begin());
+    epetra_pod_basis.Multiply(false, epetra_reduced_solution, solution);
+    this->dg->solution += reference_solution;
+    this->dg->assemble_residual();
+    epetra_pod_basis.Multiply(true, epetra_right_hand_side, epetra_reduced_rhs);
+    double new_residual;
+    epetra_reduced_rhs.Norm2(&new_residual);
+    new_residual /= this->dg->right_hand_side.size();
+
+    this->pcout << " Step length " << step_length << ". Old residual: " << initial_residual << " New residual: " << new_residual << std::endl;
+
+    int iline = 0;
+    for (iline = 0; iline < maxline && new_residual > initial_residual * reduction_tolerance_1; ++iline) {
+        step_length = step_length * step_reduction;
+        reduced_solution = old_reduced_solution;
+        reduced_solution.add(step_length, this->reduced_solution_update);
+        epetra_pod_basis.Multiply(false, epetra_reduced_solution, solution);
+        this->dg->solution += reference_solution;
+        this->dg->assemble_residual();
+        epetra_pod_basis.Multiply(true, epetra_right_hand_side, epetra_reduced_rhs);
+        epetra_reduced_rhs.Norm2(&new_residual);
+        new_residual /= this->dg->right_hand_side.size();
+        this->pcout << " Step length " << step_length << " . Old residual: " << initial_residual << " New residual: " << new_residual << std::endl;
+    }
+    this->pcout << "Full-order residual norm: " << this->dg->get_residual_l2norm() << std::endl;
 
-    pod->getPODBasis()->vmult(this->solution_update, *this->reduced_solution_update);
+    if(iline == maxline && new_residual > initial_residual){
+        this->pcout << "No decrease in residual. Maximum line search interations reached." << std::endl;
+    }
```

Comment:
Unfortunately not. The issue is for the Petrov-Galerkin ODE solver, I would need to remultiply the system matrix by the POD basis again in a separate line search function, which is quite expensive. By keeping it in the same function, I can avoid doing that. It could be done for the Galerkin solver, but as I've mentioned in other comments, I am likely going to create a parent class between them in the future, so no point in doing that right now.

Reply:
Okay, sounds good.